### PR TITLE
Add ZKStarN, ZStarNUnverified and ZStarNBuilder structs

### PR DESCRIPTION
**Pull request by Ra1issa**
This PR addresses issue #40 and adds: 

1.`ZKStarN`: struct that represents elements of the group. An instance of `ZKStarN` has a `value` that represents the group element and a modulus of the group (i.e. `builder`) passed by reference as an instance of `ZStarNBuilder`. An instance of `ZKStarN` can only be constructed after validating that  a `ZStarNUnverified` is a proper member of the group (i.e. not zero, relatively prime to the modulus, smaller than the modulus) vis a vis the modulus defined by `ZStarNBuilder` .  An instance of `ZKStarN` can be serialized, in which case only its value is serialized. 

2.`ZStarNUnverified`: struct which represents elements that have not yet been validated as proper members of the group. In order to construct an element of `ZKStarN` from its serialized value, it must first be deserialized as `ZStarNUnverified` and then validated vis a vis a `ZStarNBuilder` .

3.`ZStarNBuilder`:  which is a wrapper around a group's modulus. 

4.Tests that for serialization and for checking that elements of `ZKStarN` are properly validated.

### CODE REVIEW

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:79]: question: Why are the tests in here called `_fast`? I usually lean toward names that are closer to documentation, since I usually read them in the terminal and want to quickly know what failed. e.g. `zkstar_constructor_rejects_elements_outside_group` or `zkstar_verification_works` for the first one.

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:4]: todo: Add module-level docs here, please! We have been inconsistent about this in the rest of the library, but I like to have at least one sentence up here to show up in the built docs.

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:16]: nit: not the only way! See top-level comment :)

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:39]: suggestion: This is a circular definition ;) I've sprinkled this around in other suggestions, but we _have_ to say "multiplicative group of integers mod `N`" here, at the very least. With `N` defined by the builder.

nit: verifier -> verified.

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:46]: todo: We also need a function to generate a random element and replace `utils::random_bn_in_z_star`. I would suggest keeping the utils function for now (so you don't have to go change all the use cases); we'll plan to remove it in the second PR.

I'm flexible about whether you prefer `ZStarN::random() -> Self` or `ZStarNBuilder::random() -> ZStarN`.

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:39]: Also, document what the representation is -- I think these are always positive, right? We've been running into some issues with different modules using `[-N/2, N/2]` vs `[0, N)`. I'm not sure which one we actually want here, so just documenting what it does now should be enough to help us debug later if there is a representation problem.

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:70]: question: When do you expect this to be used? I think we should only make `ZStarNUnverified` by deserializing, and not by manual construction.

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:13]: thought: When we integrate this into the library, we might find that we are usually making this from either a `PaillierEncryptionKey` or a `RingPedersen`. We might restrict this API if that's the case (like a `from_paillier_key` and `from_ring_pedersen` method instead of generic, public `new`), but I'm not sure yet.

**Review by marsella — CHANGES_REQUESTED** [review]
Thank you for this big API! The overall shape of this is great and I think will set us up really well. I have quite a lot of comments already, so I'm gonna propose that we slap some `#[allow(unused)]`s around and merge this once the feedback is addressed, then do a separate PR to complete the issue and integrate it into the library.

Nit: Please edit your doc comments to be full sentences -- a lot of them are missing end punctuation.

General question: Does the `deserialize` macro fail for `ZStarN`? We're in kind of a weird liminal space where I want to change the entire way we do to/from bytes to use serde (and that is where the proposed thing in this issue came from), but the library currently uses those macros with no type enforcement. I don't think this will be solved in this PR but if it is a problem I'd like to document it.

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:79]: I was following some older naming conventions in `paillier.rs` where depending on whether or not you generate the primes safely or not you had a `_slow`, `_fast` suffix. I can definitely remove it.

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:70]: For the time being I was only using it in the tests. But I can totally remove it.

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:13]: I am reading this as: we will decide this during a second PR when we go ahead and start replacing all the `BigNumber` in the proofs and see what the library actually requires :)

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:49]: For the time being I was only using it in the tests. But I can totally remove it.

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:23]: Im not sure I understand what you mean ? You want to double check if the element is strictly greater than 0, rather than ust !=0 ?

Comment (marsella) [reply src/zkstar.rs:23]: Yeah, `BigNumber`s can be negative, so I think it's possible for the unverified value to be out of range because it's negative.

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:79]: Ah, I see. We just recently updated the CI to run ignored tests after you merge to main, so any test that is both slow and redundant can be tagged with `#[ignore = "whatever reason"]`. So now we can assume that anything not ignored is reasonably fast!

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:70]: Gotcha. I don't mind building the type manually in the tests (like, `let thing = ZStarUnverified(BigNumber::one())`), but if you prefer to use the `new` function, I'd say just flag it with `#[cfg(test)]`.

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:13]: Exactly.

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:49]: I'm fine with either removing it or tagging with `#[cfg(test)]`, but I would lean slightly toward removing because this doesn't seem particularly easier to call than the builder version. It's also kind of nice for the test to reflect expected external usage (although that's often not reasonable).

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:23]: Is the idea here that the element cannot be negative or that |element| < |modulus| ?

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:39]: Did you mean `(0, N)` ? 0 can't be inclusive.

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:46]: I prefer `ZStarN::random() -> Self`, because I am guessing that that's what the programmer will intend. The later one can be a bit confusing if only because the type of what its called on does not match the type that is returned :) Let me know if that sounds reasonable.

Comment (Ra1issa) [reply src/zkstar.rs:46]: Regarding the method itself: are we still thinking of NOT passing the rng ? Its usually cleaner to pass it because it makes testing and modifying it easier.

File: src/zkstar.rs
Comment (Ra1issa) [reply src/zkstar.rs:7]: Im not sure I understand what you are asking for here. Can you refer me to an example of what you'd like to see ?

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:42]: nits: add code formatting and `.`

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:49]: nits: 
- "An element in the multiplicative group..."
- code formatting on link
- `.` at the end of the first sentence
- add a newline between the two sentences (so when a text editor hovers on a `ZStarN`, it just pops up the first sentence)

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:56]: Yeah, I like that too.

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:57]: nit: backticks on link and `.`

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:66]: nits:
- `.` at end of first and second sentences
- newline between first and second sentences
- ticks on links

File: src/zkstar.rs
Comment (marsella) [inline src/zkstar.rs:98]: suggestion: Are you open to calling this something more descriptive? This is a personal pet peeve more than anything, but I don't like having `test` in the name of tests for rust because they're already flagged with `#[test]`. 

I prefer names like `zkstar_constructor_rejects_elements_outside_group` or `zkstar_verification_works`.

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs]: We discussed in person, but the element should be nonnegative and less than the modulus.

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:6]: You can try building / looking at the docs for [zkchannels-crypto](https://github.com/boltlabs-inc/libzkchannels-crypto/tree/main/zkchannels-crypto) -- the pedersen, pointcheval_sanders, and proofs modules are all good examples.

Generally speaking, I like to have one sentence that's a quick summary for anyone scanning the modules when they build the docs, and then a few more sentences with any particular context that's useful for someone coming into the library. E.g. in this case, I vetoed all your nice suggestions about adding general-purpose functionality (like a method to make a random builder) because I don't think we need it in this crate and I don't want to get confused later about why it's there. But clearly that's something any reasonable cryptographer might notice is missing when they look at the module, so I'd explain in the docs that this is _just_ the application-specific tools we need for this library.

Depending on how complex the module is, I might also add a few comments about expected usage and any major design decisions we made. For example, here we might explain the point of the builder and how it enforces correct construction and prevents silent failures from trying to combine operands of incompatible groups together. But I don't think we need to do that for this module at this point.

I might write it like:
```
//! Utilities for creating and using elements of the multiplicative group of integers mod `N`.
//!
//! This module is designed for protocol-specific use and doesn't capture all the 
//! possible functionalities on group elements.
```

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs]: Yes, thanks!

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs:84]: nit: `#[cfg(test)]` is still missing.

**Comment by marsella** [comment]
Sorry, I must have published the review before I meant to. High-level comment is: Thanks for making all these changes! I like the way this is shaping up. I have some more comments; they're all nitpicky feedback on the docs. I'm a little particular about having consistent documentation :)

File: src/errors.rs
Comment (marsella) [inline src/errors.rs:37]: suggestion: On all of these, is `ZK*` the right shorthand? I've been doing `Z*_N` in internal docs.

**Comment by marsella** [comment]
One other comment -- please plan to rebase to main and squash these changes into one commit before merging. You can do it now or after approval.

File: src/zkstar.rs
Comment (marsella) [reply src/zkstar.rs]: Love the new names, thank you!

**Comment by marsella** [comment]
Most of the links to types I see in here still don't have backticks on them. `[ZStarN]` should become ```[`ZStarN`]```, for example. I've left unresolved all the comments where that was one of the items. Everything else looks good to me, though.

**Review by marsella — APPROVED** [review]
```[`Thank you for fixing!!`]```

**Pull request merged**
